usecases/objects/references_add_test.go

```diff
@@ -13,67 +13,225 @@ package objects
-			Beacon: strfmt.URI("weaviate://localhost/d18c8e5e-a339-4c15-8af6-56b0cfe33ce7"),
-		}
 		expectedRef := &models.SingleRef{
 			Beacon: strfmt.URI("weaviate://localhost/d18c8e5e-a339-4c15-8af6-56b0cfe33ce7"),
 		}
 		expectedRefProperty := "hasAnimals"
-		vectorRepo.On("AddReference", mock.Anything, expectedRefProperty, expectedRef).Return(nil)
+		m.repo.On("AddReference", cls, mock.Anything, expectedRefProperty, expectedRef).Return(nil)
 
-		err := manager.AddObjectReference(context.Background(), nil, strfmt.UUID("my-id"), "hasAnimals", newRef)
+		err := m.AddObjectReference(context.Background(), nil, &req)
 		require.Nil(t, err)
-		vectorRepo.AssertExpectations(t)
+		m.repo.AssertExpectations(t)
 	})
+	t.Run("source object missing", func(t *testing.T) {
+		req := AddReferenceInput{
+			ID:       strfmt.UUID("my-id"),
+			Property: "hasAnimals",
+			Ref: models.SingleRef{
+				Beacon: strfmt.URI("weaviate://localhost/d18c8e5e-a339-4c15-8af6-56b0cfe33ce7"),
+			},
+		}
+		m := newFakeGetManager(zooAnimalSchemaForTest())
+		m.repo.On("ObjectByID", mock.Anything, mock.Anything, mock.Anything).Return(nil, nil)
+		err := m.AddObjectReference(context.Background(), nil, &req)
+		require.NotNil(t, err)
+		if !err.BadRequest() {
+			t.Errorf("error expected: not found error got: %v", err)
+		}
+	})
+	t.Run("source object missing", func(t *testing.T) {
+		req := AddReferenceInput{
+			ID:       strfmt.UUID("my-id"),
+			Property: "hasAnimals",
+			Ref: models.SingleRef{
+				Beacon: strfmt.URI("weaviate://localhost/d18c8e5e-a339-4c15-8af6-56b0cfe33ce7"),
+			},
+		}
+		m := newFakeGetManager(zooAnimalSchemaForTest())
+		m.repo.On("ObjectByID", mock.Anything, mock.Anything, mock.Anything).Return(nil, errors.New("any"))
+		err := m.AddObjectReference(context.Background(), nil, &req)
+		require.NotNil(t, err)
+		if err.Code != StatusInternalServerError {
+			t.Errorf("error expected: internal error, got: %v", err)
+		}
+	})
+}
+
+func Test_ReferenceAdd(t *testing.T) {
+	t.Parallel()
+	var (
+		cls         = "Zoo"
+		prop        = "hasAnimals"
+		id          = strfmt.UUID("d18c8e5e-000-0000-0000-56b0cfe33ce7")
+		refID       = strfmt.UUID("d18c8e5e-a339-4c15-8af6-56b0cfe33ce7")
+		uri         = strfmt.URI("weaviate://localhost/d18c8e5e-a339-4c15-8af6-56b0cfe33ce7")
+		anyErr      = errors.New("any")
+		ref         = models.SingleRef{Beacon: uri}
+		expectedRef = models.SingleRef{Beacon: uri}
+		req         = AddReferenceInput{
+			Class:    cls,
+			ID:       id,
+			Property: prop,
+			Ref:      ref,
+		}
+	)
+
+	tests := []struct {
+		Name string
+		// inputs
+		Req AddReferenceInput
+
+		// outputs
+		ExpectedRef models.SingleRef
+		WantCode    int
+		WantErr     error
+		SrcNotFound bool
+		// control errors
+		ErrAddRef      error
+		ErrTagetExists error
+		ErrSrcExists   error
+		ErrAuth        error
+		ErrLock        error
+		ErrSchema      error
+		// Stage: 1 -> validation(), 2 -> target exists(), 3 -> target exists(), 4 -> AddReference()
```

Comment:
```
// Stage: 1 -> validation(), 2 -> target exists(), 3 -> target exists(), 4 -> AddReference()
```

stages 1 and 2 are the same. perhaps one was intended to be `source exists()` or similar?

Reply:
You're right stage 3 should be 3 -> source exists()